src/firebase-resource-manager.ts

```diff
@@ -0,0 +1,178 @@
+  ORGANIZATION = "organization",
+  FOLDER = "folder",
+}
+
+export interface ParentResource {
+  id: string;
+  type: ParentResourceType;
+}
+
+export class FirebaseResourceManager {
+  private poller: LroPoller = new LroPoller();
+
+  async createFirebaseProject(
+    projectId: string,
+    projectDisplayName: string,
+    parentResource?: ParentResource
+  ): Promise<{ projectNumber: string; projectId: string }> {
+    const cloudProjectCreationOperation = await this.createCloudProject(
+      projectId,
+      projectDisplayName,
+      parentResource
+    );
+    const { projectNumber } = await this.pollCloudProjectCreationOperation(
+      cloudProjectCreationOperation
+    );
+    const addFirebaseOperation = await this.addFirebaseToCloudProject(projectId);
+    const projectData = await this.pollAddFirebaseToCloudProjectOperation(addFirebaseOperation);
+
+    logger.info("");
+    logger.info("🎉🎉🎉 Your Firebase project is ready! 🎉🎉🎉");
+    logger.info("");
+    logger.info("Project information:");
+    logger.info(`   - Project ID: ${clc.bold(projectData.projectId)}`);
+    logger.info(`   - Project Name: ${clc.bold(projectData.displayName)}`);
+    logger.info("");
+    logger.info("Firebase console is available at");
+    logger.info(`https://console.firebase.google.com/project/${clc.bold(projectId)}/overview`);
+    return { projectNumber, projectId };
+  }
+
+  /**
+   * @return {Promise} this function returns a promise that resolves to the resource name of the
+   *     create cloud project LRO, or rejects if an error is thrown
+   */
+  private async createCloudProject(
+    projectId: string,
+    projectDisplayName: string,
+    parentResource?: ParentResource
+  ): Promise<string> {
+    const spinner = new OraWrapper("Creating Google Cloud Platform project");
+    spinner.start();
+
+    try {
+      const response = await api.request("POST", "/v1/projects", {
+        auth: true,
+        origin: api.resourceManagerOrigin,
+        timeout: 15 * ONE_SECOND_MILLIS,
+        data: { projectId, name: projectDisplayName, parent: parentResource },
+      });
+      spinner.succeed();
+      return response.body.name; /* Operation resource name */
+    } catch (err) {
+      spinner.fail();
+      logger.debug(err.message);
+      throw new FirebaseError(
+        "Failed to create Google Cloud project. See firebase-debug.log for more info.",
+        { exit: 2, original: err }
+      );
+    }
+  }
+
+  /**
+   * @return {Promise} this function returns a promise that resolves to the new cloud project
+   *     information
+   */
+  private async pollCloudProjectCreationOperation(operationResourceName: string): Promise<any> {
+    const pollerOptions: LroPollerOptions = {
+      pollerName: "Project Creation Poller",
+      apiOrigin: api.resourceManagerOrigin,
+      apiVersion: "v1",
+      operationResourceName,
+    };
+    const spinner = new OraWrapper("Waiting for project creation to be completed");
```

Comment:
Ideally I'd like to remove this spinner and just use a single spinner for "Creating Google Cloud Platform project" which encompasses both the initial create project call and the LRO polling.

Same goes for combining the spinners for calling addFirebase and polling its LRO.

Reply:
I moved the polling logic into the private `createCloudProject()` and `addFirebaseToCloudProject()`, reducing some code duplication and 2 spinners.